### PR TITLE
feat: implement a /neetcode command displaying the NeetCode code solution for the selected problem and programming language

### PR DESCRIPTION
**Pull request by cf841**
## Changes
- Implemented the function /neetcode, which takes a programming language covered by neetcode, and a problem from leetcode, either by id, name or URL
- Get's the ID and title for the problem, from the leetcode graphql, and uses this to find the corresponding solution in the neetcode git repository. 
- generalized `ProblemSearchModal` so that different commands can call this one function and get the specific data they need from the graphql, by passing an embed function. 
- Added status checking to fetch_data
- Made `search_problem` more useful by returning title and id, rather than just title.
- Updated /help.

### CODE REVIEW

File: cogs/neetcode.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    # To prevent circular imports
+    from bot import DiscordBot
+
+class NeetcodeCog(commands.Cog):
+    
+    def __init__(self, bot:"DiscordBot") -> None:
+        self.bot = bot
+
+    @app_commands.command(name="neetcode")
+    async def neet(self, interaction: discord.Interaction) -> None:
```
Comment (Kevin-Roman) [inline cogs/neetcode.py:20]: call the function `neetcode`

File: cogs/problems.py [hunk 2]
```diff
@@ -55,6 +55,7 @@ async def random_problem(
         :param difficulty: The desired difficulty level
         """
         embed = await random_question_embed(self.bot, difficulty.value)
+
```
Comment (Kevin-Roman) [inline cogs/problems.py:58]: unnecessary change to this file

File: ui/embeds/problems.py [hunk 3]
```diff
@@ -35,6 +36,14 @@ async def search_question_embed(bot: "DiscordBot", search_text: str) -> discord.
...
     return embed
 
+# Neetcode Search by id
+async def neetcode_search_embed(bot: "DiscordBot", search_text: str) -> discord.Embed:
+    question_title, question_id = await search_question(bot, search_text)
+
+    if not question_id:
+        return question_error_embed()
+    embed = await neetcode_embed(bot, question_id, question_title)
+    return embed
```
Comment (Kevin-Roman) [inline ui/embeds/problems.py:46]: Move this to `embeds/neetcode.py`

File: ui/modals/neetcode.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    def __init__(self, bot: "DiscordBot") -> None:
+        self.bot = bot
+        super().__init__()
+
+    async def on_submit(self, interaction: discord.Interaction) -> None:
+        await interaction.response.defer()
+
+        embed = await neetcode_search_embed(self.bot, self.search_query_answer.value)
+
+        await interaction.followup.send(embed=embed)
```
Comment (Kevin-Roman) [inline ui/modals/neetcode.py:28]: Rather than creating a new NeetCodeSearchModal, similar to ProblemSearchModal, let's just make ProblemSearchModal more general. We can do this by add a new param to the constructor of ProblemSearchModal called "submit_embed_creator" or something along those lines, and store it as an attribute. You can then replace the `search_question_embed` in `on_submit` with the `self.submit_embed_creator`, and then we can just pass the appropriate embed creator when we instantiate ProblemSearchModal.

So this allows us to pass the embed creator we want (e.g. search_question_embed with neetcode_search_embed (which I'd rename to search_neetcode_embed), and only have one modal class rather than two similar ones.

File: cogs/neetcode.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        Scala = Language.SCALA
+        Dart = Language.DART    
+    
+    def __init__(self, bot:"DiscordBot") -> None:
+        self.bot = bot
+
+    @app_commands.command(name="neetcode")
+    async def neetcode(self, interaction: discord.Interaction, language: LanguageField = LanguageField.Python) -> None:
+        """
+        Get neetcode solution of leetcode propblem
```
Comment (Kevin-Roman) [inline cogs/neetcode.py:39]: *problem

File: ui/embeds/neetcode.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    # To prevent circular imports
+    from bot import DiscordBot
+
+# Neetcode Search by id
+async def search_neetcode_embed(bot: "DiscordBot", search_text: str, language: Language) -> discord.Embed:
+
+    try:
+        question_title, question_id = await search_question(bot, search_text)
+    except Exception as e:
+        return await question_error_embed()
```
Comment (Kevin-Roman) [inline ui/embeds/neetcode.py:17]: No need to catch exceptions, as ValueErrors and request related exceptions are caught within `search_question` and `bot.http_client.post_data` respectively.

Comment (Kevin-Roman) [reply ui/embeds/neetcode.py:17]: But make sure to check whether either `question_title` and `question_id` are empty, and if so, return nothing.

Comment (Kevin-Roman) [reply ui/embeds/neetcode.py:17]: question_error_embed isn't a coroutine, so no need to await.

File: utils/neetcode.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    # To prevent circular imports
+    from bot import DiscordBot
+from constants import Language
+
+async def generate_neetcode_link(bot: "DiscordBot", question_id: str, question_title: str, language: Language):
```
Comment (Kevin-Roman) [inline utils/neetcode.py:7]: There is no need for this function to be a coroutine, so remove the `async` from this function, and the `await`s used when calling this function.

File: ui/embeds/neetcode.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    return embed
+
+async def neetcode_embed(bot: "DiscordBot", question_id: str, question_title: str, language: Language) -> discord.Embed:
+
+    info = await fetch_question_info(bot, question_title)
+    neetcode_link = await generate_neetcode_link(bot, question_id, question_title, language)
+    response = await bot.http_client.fetch_data(neetcode_link, timeout=10)
+
+    if "Not Found" in response:
+        return await neetcode_error_embed()
```
Comment (Kevin-Roman) [inline ui/embeds/neetcode.py:29]: use:
`if not response_data:`

File: ui/embeds/neetcode.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        return await question_error_embed()
+    
+    embed = await neetcode_embed(bot, question_id, question_title, language)
+    return embed
+
+async def neetcode_embed(bot: "DiscordBot", question_id: str, question_title: str, language: Language) -> discord.Embed:
+
+    info = await fetch_question_info(bot, question_title)
+    neetcode_link = await generate_neetcode_link(bot, question_id, question_title, language)
+    response = await bot.http_client.fetch_data(neetcode_link, timeout=10)
```
Comment (Kevin-Roman) [inline ui/embeds/neetcode.py:26]: this returns the response data, so rename to `response_data`

File: ui/embeds/neetcode.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    )
+    return embed
+
+async def neetcode_error_embed() -> discord.Embed:
+    embed = discord.Embed(
+        title="Error",
+        description="Failed to fetch Neetcode solution, or no solution exists",
+        colour=discord.Colour.red(),
+    )
+    return embed
```
Comment (Kevin-Roman) [inline ui/embeds/neetcode.py:45]: Use the `error_embed` from `embeds/common.py`

File: utils/problems.py
Comment (Kevin-Roman) [inline utils/problems.py:1]: Everything inside this file (except `fetch_daily_question`) are your changes from your other PR, so please remove all the changes back to the original version, and make a commit. In the future, make sure to correctly change branches and stash your changes so that work isn't transferred and accidentally pushed in a commit of another branch.

File: cogs/neetcode.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        """                                                         
+        await interaction.response.send_modal(ProblemSearchModal(self.bot, search_neetcode_embed, language.value))
```
[the remark this answers is not in the document]
Comment (cf841) [reply cogs/neetcode.py:41]: Will do this tomorrow hopefully

File: utils/http_client.py [hunk 12]
```diff
@@ -31,8 +31,9 @@ async def fetch_data(self, *args, **kwargs) -> str | None:
         try:
             async with self.session.get(*args, **kwargs) as response:
                 # 404 response is expected for some requests.
-                if response.status != 404:
+                if response.status == 200:
                     return await response.text()
+            return
```
Comment (Kevin-Roman) [inline utils/http_client.py:36]: no need to return anything here

Comment (Kevin-Roman) [reply utils/http_client.py:36]: In order to keep the code consistent, could you match the code used in post_data please. Aka, use `match case` statement, and similarly log f"Fetch request error: (code: {response.status})" if the status code is not 200

Comment (Kevin-Roman) [reply utils/http_client.py:36]: In order to keep the code consistent, could you match the code used in post_data please. Aka, use `match case` statement, and similarly log a f"Fetch request error: (code: {response.status})" message if the status code is not 200

**Comment by cf841** [comment]
Cleaned up fetch_data. Only ever seen a 404 in my testing, so should cover all bases

**Comment by cf841** [comment]
- Implemented the function /neetcode, which takes a programming language covered by neetcode, and a problem from leetcode, either by id, name or URL
- Get's the ID and title for the problem, from the leetcode graphql, and uses this to find the corresponding solution in the neetcode git repository. 
- generalized `ProblemSearchModal` so that different commands can call this one function and get the specific data they need from the graphql, by passing an embed function. 
- Added status checking to fetch_data
- Made `search_problem` more useful by returning title and id, rather than just title.

File: ui/embeds/neetcode.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+) -> discord.Embed:
+
+    info = await fetch_question_info(bot, question_title)
+    neetcode_link = generate_neetcode_link(question_id, question_title, language)
+    response_data = await bot.http_client.fetch_data(neetcode_link, timeout=10)
+    if not response_data:
+        return error_embed()
+
+    embed = discord.Embed(
+        title=f"{question_id}. {info.title} - {language.value.capitalize()} Solution",
```
Comment (Kevin-Roman) [inline ui/embeds/neetcode.py:40]: I think `{question_id}. {info.title} - NeetCode Solution ({language.value.capitalize()})` would better credit NeetCode whilst also mentioning the, not as important, programming language of the solution.

**Review by Kevin-Roman — CHANGES_REQUESTED** [review]
Request a re-review once you've addressed the feedback provided.

File: ui/embeds/neetcode.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+        url=info.link,
+        description=f"```python\n{response_data}```",
```
[the remark this answers is not in the document]
Comment (cf841) [reply ui/embeds/neetcode.py:42]: oops meant to do this and forgot

File: ui/embeds/neetcode.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+        colour=discord.Colour.blue(),
```
Comment (Kevin-Roman) [inline ui/embeds/neetcode.py:43]: Make this white

Comment (cf841) [reply ui/embeds/neetcode.py]: no white available, went with next best thing

Comment (Kevin-Roman) [reply ui/embeds/neetcode.py]: looks good

**Review by Kevin-Roman — APPROVED** [review]
LGTM!
